// business-model/dist/src/entities/Role.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRoleCollection = exports.rolesCollection = exports.AuthRole = void 0;
const serverless_mongodb_utils_1 = require("serverless-mongodb-utils");
var AuthRole;
(function (AuthRole) {
    AuthRole["SuperAdmin"] = "SUPER_ADMIN";
    AuthRole["Admin"] = "ADMIN";
    AuthRole["User"] = "USER";
})(AuthRole || (exports.AuthRole = AuthRole = {}));
exports.rolesCollection = 'roles';
const createRoleCollection = () => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const database = yield (0, serverless_mongodb_utils_1.connectToDb)();
        // Create a collection
        const options = {
            validator: {
                $jsonSchema: {
                    bsonType: 'object',
                    required: ['_id', 'name'],
                    properties: {
                        _id: {
                            bsonType: 'objectId'
                        },
                        name: {
                            enum: Object.values(AuthRole),
                            bsonType: 'string'
                        }
                    }
                }
            }
        };
        yield database.createCollection(exports.rolesCollection, options);
    }
    catch (error) {
        console.error(`Error in creating Appraisal Card : ${String(error)}`);
    }
});
exports.createRoleCollection = createRoleCollection;
