// propro-auth-middleware/emails/email.controller.js
import { templateData } from "./template.data";
import NotifLocal from "./NotifLocal";
import { render } from "@react-email/render";
import FormData from "form-data";
import Mailgun from "mailgun.js";

const email = async (emailAddress, app, template, ...args) => {
  // const { email: emailAddress, app } = JSON.parse(req.body);

  // const template = req.query.template;

  const mailgun = new Mailgun(FormData);

  const client = mailgun.client({
    username: "email-service",
    key: process.env.MAILGUN_API_KEY,
    url: "https://api.eu.mailgun.net",
  });

  const selectedTemplate = templateData({ ...args })[template];

  try {
    const emailHTML = render(
      <NotifLocal
        body={selectedTemplate.body}
        app={app}
        label={selectedTemplate.button?.label}
        url={selectedTemplate.button?.url}
      />,
    );

    const message = {
      from: process.env.EMAIL_FROM,
      to: emailAddress,
      subject: selectedTemplate.subject,
      html: emailHTML,
    };

    const result = await client.messages.create(
      process.env.EMAIL_DOMAIN,
      message,
    );
    res.send(result);
  } catch (error) {
    res.end(JSON.stringify(error));
  }
};

export default email;
