// propro-auth-middleware/emails/NotifLocal.tsx
import { Button } from "@react-email/button";
import { Container } from "@react-email/container";
import { Head } from "@react-email/head";
import { Html } from "@react-email/html";
import { Img } from "@react-email/img";
import { Section } from "@react-email/section";
import { Text } from "@react-email/text";
import React from "react";

const NotifLocal = ({
  body,
  app,
  label,
  url,
}: {
  body: string;
  app: string;
  label?: string;
  url?: string;
}) => {
  return (
    <Html>
      <Head />
      <Section style={main}>
        <Container style={container}>
          <Img src={getLogoUrl(app)} />
          <Section>
            <Text style={text}>
              {body ||
                "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed tempus quam vehicula, accumsan orci a, iaculis nulla. Sed a velit vel augue auctor vehicula. Phasellus interdum aliquet mattis. "}
            </Text>
            {label && (
              <Button pX={20} pY={12} style={btn} href={url}>
                {label || "temp label"}
              </Button>
            )}
          </Section>
        </Container>
      </Section>
    </Html>
  );
};

const getLogoUrl = (app: string) => {
  switch (app) {
    case "hubhub":
      return "/assets/logos/hubhub.svg";
    case "propro":
      return "/assets/logos/propro.svg";
    default:
      return "/assets/logos/hubhub.svg";
  }
};

const main = {
  backgroundColor: "#ffffff",
  margin: "0 auto",
};

const container = {
  border: "1px solid #eaeaea",
  borderRadius: "5px",
  margin: "40px auto",
  padding: "20px",
  width: "465px",
};

const h1 = {
  color: "#000",
  fontFamily:
    "-apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif",
  fontSize: "24px",
  fontWeight: "normal",
  textAlign: "center" as const,
  margin: "30px 0",
  padding: "0",
};

const text = {
  color: "#000",
  fontFamily:
    "-apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif",
  fontSize: "14px",
  lineHeight: "24px",
};

const btn = {
  backgroundColor: "#000",
  borderRadius: "5px",
  color: "#fff",
  fontFamily:
    "-apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif",
  fontSize: "12px",
  fontWeight: 500,
  lineHeight: "50px",
  textDecoration: "none",
  textAlign: "center" as const,
};

export default NotifLocal;
