// propro-auth-middleware/emails/template.data.js
export const templateData = ({ ...args }) => ({
  "new-account": {
    subject: "Welcome to ProPro",
    body: "Thank you for signing up to ProPro. Please click the link below to verify your email address.",
    button: {
      label: "Verify Your Email",
      url: "{{domain}}?token={{content.verification_token}}",
    },
  },
  "duplicate-user": {
    subject: "Important! Your ProPro Account",
    body: "We noticed you already have a registered account with ProPro. To prevent confusion, your original password remains unchanged.",
    button: {
      label: "Sign In",
      url: "{{domain}}/signin",
    },
  },
  new_plan: {
    subject: "Your ProPro Plan",
    body: "Thank you for choosing the {{content.plan}} plan. Your card has been successfully charged {{content.price}}.",
    button: {
      label: "Your Dashboard",
      url: "{{domain}}/dashboard",
    },
  },
  "new-user": {
    subject: "Welcome to ProPro",
    body: "Thank you for signing up to ProPro. Please click the link below to verify your email address.",
    button: {
      label: "Verify Your Email",
      url: "{{domain}}?token={{content.verification_token}}",
    },
  },
  "plan-updated": {
    subject: "Your Billing Plan Updated",
    body: "Confirmation: Your billing plan has been updated to the {{content.plan}} plan.",
    button: {
      label: "Manage Your Plan",
      url: "{{domain}}/account/billing",
    },
  },
  "card-updated": {
    subject: "Credit Card Information Updated",
    body: "Confirmation: Your credit card information has been updated. Contact us immediately if you did not authorize this change.",
    button: {
      label: "Contact Support",
      url: "mailto:support@yourdomain.com",
    },
  },
  "account-closed": {
    subject: "Account Closure Confirmation",
    body: "Your account has been closed and all associated data has been removed from our servers. You're welcome to rejoin ProPro anytime.",
    button: {
      label: "Sign Up",
      url: "{{domain}}/signup",
    },
  },
  "invite-accepted": {
    subject: "Team Invitation Accepted",
    body: "Great news! {{content.friend}} has accepted your invitation to join your ProPro team.",
    button: {
      label: "Invite More Friends",
      url: "{{domain}}/account/users",
    },
  },
  "password-updated": {
    subject: "Password Change Confirmation",
    body: "Your ProPro password has been changed. If you did not initiate this change, please contact our support team immediately.",
    button: {
      label: "Contact Support",
      url: "mailto:support@yourdomain.com",
    },
  },
  "password-reset": {
    subject: "Password Reset Request",
    title: "Forgot Your ProPro Password?",
    body: "To reset your password, click the link below.",
    button: {
      label: "Reset Password",
      url: "{{domain}}?token={{content.token}}",
    },
  },
  invite: {
    subject: "Team Invitation from a Friend",
    body: "You've been invited by {{content.friend}} to join their ProPro team. Sign up now to collaborate.",
    button: {
      label: "Sign Up Now",
      url: "{{domain}}?id={{content.id}}&email={{content.email}}",
    },
  },
  contact: {
    subject: "New Contact Form Submission",
    body: "You've received a new message from {{content.name}}: <br><br> {{content.message}}",
    button: {
      label: "Respond Now",
      url: "mailto:{{content.email}}",
    },
  },
  feedback: {
    subject: "New App Feedback Received",
    body: "New feedback for your ProPro app: <br><br> Rating: {{content.rating}} <br><br>{{content.comment}}",
    button: {
      label: "View Now",
      url: "{{content.url}}",
    },
  },
  magic_signin: {
    subject: "Your ProPro Magic Sign-In Link",
    body: "You requested a magic sign-in link for ProPro. Click below to sign in without a password.",
    button: {
      label: "Sign In",
      url: "{{domain}}?token={{content.token}}",
    },
  },
  new_signin: {
    subject: "New Sign-In Alert for ProPro",
    body: "A new sign-in to your ProPro account was detected: \n\n{{content.browser}} on {{content.device}}\n\nIP: {{content.ip}}\n\nTime: {{content.time}}\n\nIf this was you, no action is needed. If not, please secure your account.",
    button: {
      label: "Secure My Account",
      url: "{{domain}}/account/security",
    },
  },
  blocked_signin: {
    subject: "Blocked Suspicious Sign-In Attempt",
    body: "A suspicious sign-in attempt to your ProPro account was blocked. If this was you, use the link below to sign in securely.",
    button: {
      label: "Sign In Securely",
      url: "{{domain}}?token={{content.token}}",
    },
  },
  help: {
    subject: "New Support Request Received",
    body: "A new support request has been submitted by {{content.name}}: <br><br> {{content.message}}",
    button: {
      label: "Respond Now",
      url: "mailto:{{content.email}}",
    },
  },
  "new-log": {
    subject: "New Log Recorded in ProPro",
    body: "A new log has been recorded in your ProPro application. Check Mission Control for details.",
    button: {
      label: "View Log",
      url: "{{domain}}/missioncontrol/logs/{{content.id}}",
    },
  },
  email_verification: {
    subject: "Verify Your ProPro Email",
    body: "Thanks for signing up to ProPro. Click the link below to verify your email address.",
    button: {
      label: "Verify Your Email",
      url: "{{domain}}?token={{content.verification_token}}",
    },
  },
  email_verification_complete: {
    subject: "Email Verified - Welcome to ProPro",
    body: "Your email address has been verified. Your ProPro account is now active.",
    button: {
      label: "Open Dashboard",
      url: "{{domain}}/dashboard",
    },
  },
  "new-email": {
    subject: "Email Address Change Confirmation",
    body: "You have changed your email address for your ProPro account. If you did not authorize this change, please contact us immediately.",
    button: {
      label: "Contact Support",
      url: "mailto:support@yourdomain.com",
    },
  },
  otp_request: {
    subject: "Your ProPro One-Time Password",
    body: "Your ProPro one-time password (OTP) is: {{content.otp}}",
    additionalInfo: "This password will expire in 10 minutes.",
  },
  subscription_expiry: {
    subject: "Your ProPro Subscription is Expiring",
    body: "Your ProPro subscription will expire in {{content.days}} days. Renew now to continue enjoying our services.",
    button: {
      label: "Renew Subscription",
      url: "{{domain}}/account/billing",
    },
  },
  payment_failure: {
    subject: "Your ProPro Payment Failed",
    body: "We were unable to process your payment. Please update your billing information to avoid interruption of your ProPro services.",
    button: {
      label: "Update Billing Information",
      url: "{{domain}}/account/billing",
    },
  },
  account_recovery: {
    subject: "ProPro Account Recovery",
    body: "To recover your ProPro account, please click the link below.",
    button: {
      label: "Recover Account",
      url: "{{domain}}?token={{content.token}}",
    },
  },
  account_verified: {
    subject: "Your ProPro Account is Verified",
    body: "Your ProPro account is now verified. You can now use all the features available.",
    button: {
      label: "Get Started",
      url: "{{domain}}/dashboard",
    },
  },
  account_verification_failed: {
    subject: "ProPro Account Verification Failed",
    body: "Your attempt to verify your ProPro account failed. Please try again or contact support if you continue to have issues.",
    button: {
      label: "Contact Support",
      url: "mailto:support@yourdomain.com",
    },
  },
  report_ready: {
    subject: "Your ProPro Report is Ready",
    body: "Your requested report from ProPro is now ready for download.",
    button: {
      label: "Download Report",
      url: "{{domain}}/report/{{content.id}}",
    },
  },
  welcome_message: {
    subject: "Welcome to ProPro!",
    body: "We're excited to have you on board. Discover the possibilities and features available with ProPro.",
    button: {
      label: "Get Started",
      url: "{{domain}}/dashboard",
    },
  },
  congratulations: {
    subject: "Congratulations from ProPro",
    body: "Congratulations on reaching a new milestone with ProPro!",
    button: {
      label: "Celebrate Now",
      url: "{{domain}}/rewards",
    },
  },
  newsletter: {
    subject: "ProPro Newsletter - {{content.month}}",
    body: "Catch up on the latest news and updates from ProPro in our monthly newsletter.",
    button: {
      label: "Read Now",
      url: "{{domain}}/newsletter/{{content.id}}",
    },
  },
  security_alert: {
    subject: "Security Alert: New Device Signed in to ProPro",
    body: "A new device has just signed into your ProPro account. If this was you, you can safely ignore this email. If it wasn't you, please secure your account.",
    button: {
      label: "Secure My Account",
      url: "{{domain}}/account/security",
    },
  },
  promotion: {
    subject: "Special Promotion Just for You from ProPro",
    body: "We have a special promotion just for you. Unlock your exclusive deal from ProPro now.",
    button: {
      label: "Unlock Now",
      url: "{{domain}}/promotions",
    },
  },
  thank_you: {
    subject: "Thank You for Choosing ProPro",
    body: "Thank you for being a valued member of the ProPro community. We're glad to have you with us.",
    button: {
      label: "Explore More",
      url: "{{domain}}/dashboard",
    },
  },
  goodbye: {
    subject: "Goodbye from ProPro",
    body: "We're sad to see you go. If you ever decide to come back, ProPro will be here waiting for you.",
    button: {
      label: "Come Back",
      url: "{{domain}}/signup",
    },
  },
  confirmation: {
    subject: "Action Required: Confirm Your Action",
    body: "Please confirm your action by clicking the button below.",
    button: {
      label: "Confirm",
      url: "{{domain}}?token={{content.token}}",
    },
  },
});
